File: app/src/main/java/com/merino/ddfilms/adapters/MovieAdapter.java
package com.merino.ddfilms.adapters;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityOptionsCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.merino.ddfilms.R;
import com.merino.ddfilms.model.Movie;
import com.merino.ddfilms.ui.MovieDetailActivity;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import lombok.Getter;
import lombok.Setter;

public class MovieAdapter extends RecyclerView.Adapter<MovieAdapter.MovieViewHolder> {
    @Setter
    @Getter
    private List<Movie> movies = new ArrayList<>();
    @Setter
    @Getter
    private List<Integer> moviesIdList = new ArrayList<>();
    @Setter
    @Getter
    private List<String> usersList = new ArrayList<>();
    private boolean isEditMode = false;
    private boolean isAddMode = false;
    private OnItemLongClickListener longClickListener;

    private OnDeleteClickListener deleteClickListener;
    private OnAddClickListener addClickListener;
    private OnCheckClickListener checkClickListener;
    private OnAddedByChangedListener addedByChangedListener;

    @Setter
    private String listID;

    public interface OnItemLongClickListener {
        boolean onItemLongClick(int position);
    }

    public interface OnDeleteClickListener {
        void onDeleteClick(int position, Movie movie);
    }

    public interface OnAddClickListener {
        void onAddClick(int position, Movie movie);
    }

    public interface OnCheckClickListener {
        void onCheckClick(int position, Movie movie);
    }

    public interface OnAddedByChangedListener {
        void onAddedByChanged(int position, Movie movie, String newAddedBy);
    }

    public void setOnItemLongClickListener(OnItemLongClickListener listener) {
        this.longClickListener = listener;
    }

    public void setOnDeleteClickListener(OnDeleteClickListener listener) {
        this.deleteClickListener = listener;
    }

    public void setOnAddClickListener(OnAddClickListener listener) {
        this.addClickListener = listener;
    }

    public void setOnCheckClickListener(OnCheckClickListener listener) {
        this.checkClickListener = listener;
    }

    public void setOnAddedByChangedListener(OnAddedByChangedListener listener) {
        this.addedByChangedListener = listener;
    }

    public void setEditMode(boolean editMode) {
        isEditMode = editMode;
        notifyDataSetChanged();
    }

    public void setAddMode(boolean addMode) {
        isAddMode = addMode;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public MovieViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_movie, parent, false);
        return new MovieViewHolder(view);
    }

    @SuppressLint("NotifyDataSetChanged")
    public void setMovies(List<Movie> movies) {
        Map<String, List<Movie>> moviesByDate = new LinkedHashMap<>();

        for (Movie movie : movies) {
            if (movie.getCreatedAt() != null) {
                moviesByDate.computeIfAbsent(movie.getCreatedAt(), k -> new ArrayList<>()).add(movie);
            }
        }

        if (!moviesByDate.isEmpty()) {
            List<Movie> sortedMovies = new ArrayList<>();
            for (Map.Entry<String, List<Movie>> entry : moviesByDate.entrySet()) {
                sortedMovies.addAll(entry.getValue());
            }
            this.movies = sortedMovies;
        } else {
            this.movies = movies;
        }
        notifyDataSetChanged();
    }

    @Override
    public void onBindViewHolder(@NonNull MovieViewHolder holder, int position) {
        Movie movie = movies.get(position);
        holder.bind(movie);

        // Determine if we should show the createdAt TextView
        if (movie.getCreatedAt() != null && !isAddMode && (position == 0 || !Objects.equals(movies.get(position - 1).getCreatedAt(), movie.getCreatedAt()))) {
            holder.createdAtTextView.setText("Añadido el " + movie.getCreatedAt());
            holder.createdAtTextView.setVisibility(View.VISIBLE);
        } else {
            holder.createdAtTextView.setVisibility(View.GONE);
        }
    }

    @Override
    public int getItemCount() {
        return movies.size();
    }

    public void removeMovie(int position) {
        movies.remove(position);
        notifyItemRemoved(position);
    }

    class MovieViewHolder extends RecyclerView.ViewHolder {
        private final ImageView posterImageView;
        private final TextView titleTextView;
        private final TextView overviewTextView;
        private final TextView yearTextView;
        private final TextView voteAverageTextView;
        private final Spinner addedBySpinner;
        private final TextView createdAtTextView;
        private final ImageButton deleteButton;
        private final ImageButton addButton;
        private final ImageButton checkButton;
        private ArrayAdapter<String> spinnerAdapter;
        private String currentAddedBy = "";

        public MovieViewHolder(@NonNull View itemView) {
            super(itemView);
            posterImageView = itemView.findViewById(R.id.poster_image_view);
            titleTextView = itemView.findViewById(R.id.title_text_view);
            overviewTextView = itemView.findViewById(R.id.overview_text_view);
            yearTextView = itemView.findViewById(R.id.year_text_view);
            voteAverageTextView = itemView.findViewById(R.id.vote_average_text_view);
            addedBySpinner = itemView.findViewById(R.id.added_by_spinner);
            createdAtTextView = itemView.findViewById(R.id.created_at_text_view);
            deleteButton = itemView.findViewById(R.id.delete_button);
            addButton = itemView.findViewById(R.id.add_button);
            checkButton = itemView.findViewById(R.id.check_button);

            setupClickListeners();
            setupSpinner();
        }

        private void setupSpinner() {
            // Crear adapter personalizado vacío inicialmente
            spinnerAdapter = new ArrayAdapter<String>(itemView.getContext(), R.layout.spinner_item_layout, new ArrayList<>()) {
                @Override
                public View getView(int position, View convertView, ViewGroup parent) {
                    View view = super.getView(position, convertView, parent);
                    TextView textView = view.findViewById(R.id.spinner_text);
                    textView.setText(getItem(position));
                    return view;
                }

                @Override
                public View getDropDownView(int position, View convertView, ViewGroup parent) {
                    if (convertView == null) {
                        convertView = LayoutInflater.from(itemView.getContext()).inflate(R.layout.spinner_dropdown_item, parent, false);
                    }
                    TextView textView = convertView.findViewById(R.id.dropdown_text);
                    String item = getItem(position);
                    textView.setText(item);

                    // Cambiar color según si está seleccionado o no
                    if (Objects.equals(item, currentAddedBy)) {
                        textView.setTextColor(itemView.getContext().getResources().getColor(R.color.gold_dark));
                    } else {
                        textView.setTextColor(itemView.getContext().getResources().getColor(android.R.color.white));
                    }

                    return convertView;
                }
            };

            addedBySpinner.setAdapter(spinnerAdapter);
        }

        private void setupSpinnerListener() {
            // Listener para detectar cambios
            addedBySpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
                @Override
                public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                    String selectedUser = spinnerAdapter.getItem(position);

                    // Solo notificar si realmente cambió el valor
                    if (addedByChangedListener != null && !Objects.equals(currentAddedBy, selectedUser)) {
                        Movie movie = getMovieAtPosition(getBindingAdapterPosition());
                        if (movie != null) {
                            // Actualizar el valor actual para futuras comparaciones
                            currentAddedBy = selectedUser;
                            addedByChangedListener.onAddedByChanged(getBindingAdapterPosition(), movie, selectedUser);
                        }
                    }
                }

                @Override
                public void onNothingSelected(AdapterView<?> parent) {
                    // No hacer nada
                }
            });
        }

        private void setupClickListeners() {
            itemView.setOnClickListener(v -> {
                if (!isEditMode) {
                    Movie movie = getMovieAtPosition(getAdapterPosition());
                    if (movie != null) {
                        Intent intent = new Intent(itemView.getContext(), MovieDetailActivity.class);
                        intent.putExtra("movie", movie);

                        ActivityOptionsCompat options = ActivityOptionsCompat.makeSceneTransitionAnimation((Activity) itemView.getContext(), posterImageView, "moviePosterTransition");

                        itemView.getContext().startActivity(intent, options.toBundle());
                    }
                }
            });

            itemView.setOnLongClickListener(v -> {
                if (longClickListener != null) {
                    return longClickListener.onItemLongClick(getBindingAdapterPosition());
                }
                return false;
            });

            deleteButton.setOnClickListener(v -> {
                int position = getBindingAdapterPosition();
                if (position != RecyclerView.NO_POSITION) {
                    deleteClickListener.onDeleteClick(position, movies.get(position));
                }
            });

            addButton.setOnClickListener(v -> {
                int position = getBindingAdapterPosition();
                if (position != RecyclerView.NO_POSITION) {
                    addClickListener.onAddClick(position, movies.get(position));
                }
            });

            checkButton.setOnClickListener(v -> {
                int position = getBindingAdapterPosition();
                if (position != RecyclerView.NO_POSITION) {
                    checkClickListener.onCheckClick(position, movies.get(position));
                }
            });
        }

        @SuppressLint("DefaultLocale")
        public void bind(Movie movie) {
            Glide.with(itemView.getContext()).load("https://image.tmdb.org/t/p/w500" + movie.getPosterPath()).into(posterImageView);
            titleTextView.setText(movie.getTitle());
            yearTextView.setText("(" + movie.getReleaseDate() + ")");
            overviewTextView.setText(movie.getOverview());
            voteAverageTextView.setText(String.format("%.1f", movie.getVoteAverage()));

            boolean shouldShowSpinner = !isAddMode && movie.getAddedBy() != null && !movie.getAddedBy().isEmpty();

            if (shouldShowSpinner) {
                addedBySpinner.setOnItemSelectedListener(null);

                List<String> options = new ArrayList<>();

                if (usersList != null && !usersList.isEmpty()) {
                    for (String user : usersList) {
                        if (!options.contains(user)) {
                            options.add(user);
                        }
                    }
                    options.add("Común"); // Dejamos siempre esta última opción
                }

                // Actualizar adapter del spinner
                spinnerAdapter.clear();
                spinnerAdapter.addAll(options);
                spinnerAdapter.notifyDataSetChanged();

                String movieAddedBy = movie.getAddedBy();
                currentAddedBy = movieAddedBy;

                int selectedPosition = options.indexOf(movieAddedBy);
                if (selectedPosition >= 0) {
                    addedBySpinner.setSelection(selectedPosition);
                }

                setupSpinnerListener();

                addedBySpinner.setVisibility(View.VISIBLE);
                posterImageView.setMaxHeight(500);
                posterImageView.setMinimumHeight(500);
            } else {
                addedBySpinner.setVisibility(View.GONE);
                addedBySpinner.setOnItemSelectedListener(null);
            }

            deleteButton.setVisibility(isEditMode ? View.VISIBLE : View.GONE);

            if (isAddMode) {
                boolean isAlreadyAdded = isMovieAlreadyAdded(movie, moviesIdList);
                addButton.setVisibility(isAlreadyAdded ? View.GONE : View.VISIBLE);
                checkButton.setVisibility(isAlreadyAdded ? View.VISIBLE : View.GONE);
            } else {
                addButton.setVisibility(View.GONE);
                checkButton.setVisibility(View.GONE);
            }
        }
    }

    public Movie getMovieAtPosition(int position) {
        if (position >= 0 && position < movies.size()) {
            return movies.get(position);
        }
        return null;
    }

    private boolean isMovieAlreadyAdded(Movie movie, List<Integer> moviesIdList) {
        for (Integer id : moviesIdList) {
            if (Objects.equals(id, movie.getId())) {
                return true;
            }
        }
        return false;
    }
}

File: app/src/main/java/com/merino/ddfilms/adapters/MovieAdapter.kt
package com.merino.ddfilms.adapters

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.Spinner
import android.widget.TextView
import androidx.core.app.ActivityOptionsCompat
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.merino.ddfilms.R
import com.merino.ddfilms.model.Movie
import com.merino.ddfilms.ui.MovieDetailActivity
import java.util.ArrayList
import java.util.LinkedHashMap

class MovieAdapter : RecyclerView.Adapter<MovieAdapter.MovieViewHolder>() {

    var movies: MutableList<Movie> = ArrayList()
        private set

    var moviesIdList: List<Int> = ArrayList()
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    var usersList: List<String> = ArrayList()
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    var isEditMode = false
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    var isAddMode = false
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    private var longClickListener: OnItemLongClickListener? = null
    private var lastClickTime: Long = 0
    private val CLICK_INTERVAL: Long = 1200

    private var deleteClickListener: OnDeleteClickListener? = null
    private var addClickListener: OnAddClickListener? = null
    private var checkClickListener: OnCheckClickListener? = null
    private var addedByChangedListener: OnAddedByChangedListener? = null

    var listID: String? = null

    fun interface OnItemLongClickListener {
        fun onItemLongClick(position: Int): Boolean
    }

    fun interface OnDeleteClickListener {
        fun onDeleteClick(position: Int, movie: Movie)
    }

    fun interface OnAddClickListener {
        fun onAddClick(position: Int, movie: Movie)
    }

    fun interface OnCheckClickListener {
        fun onCheckClick(position: Int, movie: Movie)
    }

    fun interface OnAddedByChangedListener {
        fun onAddedByChanged(position: Int, movie: Movie, newAddedBy: String)
    }

    fun setOnItemLongClickListener(listener: OnItemLongClickListener?) {
        this.longClickListener = listener
    }

    fun setOnDeleteClickListener(listener: OnDeleteClickListener?) {
        this.deleteClickListener = listener
    }

    fun setOnAddClickListener(listener: OnAddClickListener?) {
        this.addClickListener = listener
    }

    fun setOnCheckClickListener(listener: OnCheckClickListener?) {
        this.checkClickListener = listener
    }

    fun setOnAddedByChangedListener(listener: OnAddedByChangedListener?) {
        this.addedByChangedListener = listener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_movie, parent, false)
        return MovieViewHolder(view)
    }

    fun setMovies(movies: List<Movie>) {
        val moviesByDate = LinkedHashMap<String, MutableList<Movie>>()

        for (movie in movies) {
            val date = movie.createdAt
            if (date != null) {
                moviesByDate.getOrPut(date) { ArrayList() }.add(movie)
            }
        }

        if (moviesByDate.isNotEmpty()) {
            val sortedMovies = ArrayList<Movie>()
            for (entry in moviesByDate.entries) {
                sortedMovies.addAll(entry.value)
            }
            this.movies = sortedMovies
        } else {
            this.movies = ArrayList(movies)
        }
        notifyDataSetChanged()
    }

    override fun onBindViewHolder(holder: MovieViewHolder, position: Int) {
        val movie = movies[position]
        holder.bind(movie)

        // Determine if we should show the createdAt TextView
        val createdAt = movie.createdAt
        if (createdAt != null && !isAddMode && (position == 0 || movies[position - 1].createdAt != createdAt)) {
            holder.createdAtTextView.text = holder.itemView.context.getString(R.string.added_on, createdAt)
            holder.createdAtTextView.visibility = View.VISIBLE
        } else {
            holder.createdAtTextView.visibility = View.GONE
        }
    }

    override fun getItemCount(): Int = movies.size

    fun removeMovie(position: Int) {
        if (position in movies.indices) {
            movies.removeAt(position)
            notifyItemRemoved(position)
        }
    }

    inner class MovieViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val posterImageView: ImageView = itemView.findViewById(R.id.poster_image_view)
        val titleTextView: TextView = itemView.findViewById(R.id.title_text_view)
        val overviewTextView: TextView = itemView.findViewById(R.id.overview_text_view)
        val yearTextView: TextView = itemView.findViewById(R.id.year_text_view)
        val voteAverageTextView: TextView = itemView.findViewById(R.id.vote_average_text_view)
        val addedBySpinner: Spinner = itemView.findViewById(R.id.added_by_spinner)
        val createdAtTextView: TextView = itemView.findViewById(R.id.created_at_text_view)
        val deleteButton: ImageButton = itemView.findViewById(R.id.delete_button)
        val addButton: ImageButton = itemView.findViewById(R.id.add_button)
        val checkButton: ImageButton = itemView.findViewById(R.id.check_button)
        private lateinit var spinnerAdapter: ArrayAdapter<String>
        private var currentAddedBy = ""

        init {
            setupClickListeners()
            setupSpinner()
        }

        private fun setupSpinner() {
            spinnerAdapter = object : ArrayAdapter<String>(itemView.context, R.layout.spinner_item_layout, ArrayList()) {
                override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                    val view = convertView ?: LayoutInflater.from(context).inflate(R.layout.spinner_item_layout, parent, false)
                    val textView = view.findViewById<TextView>(R.id.spinner_text)
                    textView.text = getItem(position)
                    return view
                }

                override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup): View {
                    val view = convertView ?: LayoutInflater.from(context).inflate(R.layout.spinner_dropdown_item, parent, false)
                    val textView = view.findViewById<TextView>(R.id.dropdown_text)
                    val item = getItem(position)
                    textView.text = item

                    if (item == currentAddedBy) {
                        textView.setTextColor(context.resources.getColor(R.color.gold_dark, null))
                    } else {
                        textView.setTextColor(context.resources.getColor(android.R.color.white, null))
                    }

                    return view
                }
            }

            addedBySpinner.adapter = spinnerAdapter
        }

        private fun setupSpinnerListener() {
            addedBySpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                    val selectedUser = spinnerAdapter.getItem(position) ?: return

                    if (addedByChangedListener != null && currentAddedBy != selectedUser) {
                        val movie = getMovieAtPosition(bindingAdapterPosition)
                        if (movie != null) {
                            currentAddedBy = selectedUser
                            addedByChangedListener?.onAddedByChanged(bindingAdapterPosition, movie, selectedUser)
                        }
                    }
                }

                override fun onNothingSelected(parent: AdapterView<*>?) {}
            }
        }

        private fun setupClickListeners() {
            itemView.setOnClickListener {
                if (!isEditMode) {
                    val currentTime = android.os.SystemClock.elapsedRealtime()
                    if (currentTime - lastClickTime < CLICK_INTERVAL) {
                        return@setOnClickListener
                    }
                    lastClickTime = currentTime

                    val movie = getMovieAtPosition(bindingAdapterPosition)
                    if (movie != null) {
                        val intent = Intent(itemView.context, MovieDetailActivity::class.java)
                        intent.putExtra("movie", movie)

                        val options = ActivityOptionsCompat.makeSceneTransitionAnimation(
                            itemView.context as Activity,
                            posterImageView,
                            "moviePosterTransition"
                        )

                        itemView.context.startActivity(intent, options.toBundle())
                    }
                }
            }

            itemView.setOnLongClickListener {
                longClickListener?.onItemLongClick(bindingAdapterPosition) ?: false
            }

            deleteButton.setOnClickListener {
                val position = bindingAdapterPosition
                if (position != RecyclerView.NO_POSITION) {
                    deleteClickListener?.onDeleteClick(position, movies[position])
                }
            }

            addButton.setOnClickListener {
                val position = bindingAdapterPosition
                if (position != RecyclerView.NO_POSITION) {
                    addClickListener?.onAddClick(position, movies[position])
                }
            }

            checkButton.setOnClickListener {
                val position = bindingAdapterPosition
                if (position != RecyclerView.NO_POSITION) {
                    checkClickListener?.onCheckClick(position, movies[position])
                }
            }
        }

        @SuppressLint("DefaultLocale")
        fun bind(movie: Movie) {
            Glide.with(itemView.context).load("https://image.tmdb.org/t/p/w500" + movie.posterPath).into(posterImageView)
            titleTextView.text = movie.title
            yearTextView.text = "(${movie.releaseDate})"
            overviewTextView.text = movie.overview
            voteAverageTextView.text = String.format("%.1f", movie.voteAverage)

            val shouldShowSpinner = !isAddMode && !movie.addedBy.isNullOrEmpty()

            if (shouldShowSpinner) {
                addedBySpinner.onItemSelectedListener = null

                val options = ArrayList<String>()

                if (usersList.isNotEmpty()) {
                    for (user in usersList) {
                        if (!options.contains(user)) {
                            options.add(user)
                        }
                    }
                    options.add("Común")
                } else {
                    val movieAddedBy = movie.addedBy
                    if (!movieAddedBy.isNullOrEmpty()) {
                        options.add(movieAddedBy)
                    }
                    options.add("Común")
                }

                spinnerAdapter.clear()
                spinnerAdapter.addAll(options)
                spinnerAdapter.notifyDataSetChanged()

                val movieAddedBy = movie.addedBy ?: ""
                currentAddedBy = movieAddedBy

                val selectedPos = options.indexOf(movieAddedBy)
                if (selectedPos >= 0) {
                    addedBySpinner.setSelection(selectedPos)
                }

                setupSpinnerListener()

                addedBySpinner.visibility = View.VISIBLE
                posterImageView.maxHeight = 500
                posterImageView.minimumHeight = 500
            } else {
                addedBySpinner.visibility = View.GONE
                addedBySpinner.onItemSelectedListener = null
            }

            deleteButton.visibility = if (isEditMode) View.VISIBLE else View.GONE

            if (isAddMode) {
                val isAlreadyAdded = isMovieAlreadyAdded(movie, moviesIdList)
                addButton.visibility = if (isAlreadyAdded) View.GONE else View.VISIBLE
                checkButton.visibility = if (isAlreadyAdded) View.VISIBLE else View.GONE
            } else {
                addButton.visibility = View.GONE
                checkButton.visibility = View.GONE
            }
        }
    }

    fun getMovieAtPosition(position: Int): Movie? {
        return if (position in movies.indices) movies[position] else null
    }

    private fun isMovieAlreadyAdded(movie: Movie, moviesIdList: List<Int>): Boolean {
        return moviesIdList.contains(movie.id)
    }
}
